src/mo_frontend/typing.ml

```diff
@@ -972,6 +972,12 @@ and infer_exp'' env exp : T.typ =
       ot := t
     end;
     T.text
+  | ToCandidE (ot, exps) ->
+    let ts = List.map (infer_exp env) exps in
+    ot := T.Tup ts;
```

Comment:
You need to also check the type is `shared`.

Reply:
And only check infer the types of subexpressions when `not env.pre` (see ShowE above for an example).